lib/pymedphys/tests/dicom/test_connect.py

```diff
@@ -360,68 +385,34 @@ def test_dicom_sender_cli(test_dataset):
         send_directory = test_directory.joinpath("send")
         send_directory.mkdir()
         send_file = send_directory.joinpath("test.dcm")
-        test_dataset.save_as(send_file)
-
-        # Need to write the preamble or else we pydicom throws an error when reading the
-        # file... This solution is from: https://github.com/pydicom/pydicom/issues/340
-        # TODO: Not entirely sure why this is needed. Is there a cleaner solution to
-        # this?
-        fp = BytesIO()
-        fp.write(b"\x00" * 128)
-        fp.write(b"DICM")
-        f = open(send_file, "rb")
-        fp.write(f.read())
-        f.close()
-        fp.seek(0)
-        ds = pydicom.read_file(fp)
-        ds.save_as(send_file)
+        test_dataset.save_as(send_file, write_like_original=False)
 
         receive_directory = test_directory.joinpath("receive")
         receive_directory.mkdir()
 
-        listener_command = [
-            pmp_test_utils.get_executable_even_when_embedded(),
-            "-m",
-            "pymedphys",
-            "dicom",
-            "listen",
-            str(TEST_PORT),
-            "-d",
-            str(receive_directory),
-            "-a",
-            str(scp_ae_title),
-        ]
-
-        sender_command = [
-            pmp_test_utils.get_executable_even_when_embedded(),
-            "-m",
-            "pymedphys",
-            "dicom",
-            "send",
-            "-a",
-            str(scp_ae_title),
-            "localhost",
-            str(TEST_PORT),
-            str(send_file),
-        ]
-
-        # TODO: I don't love having to sleep in unit tests, but the CLI seems to need it
-        with process(listener_command):
-            time.sleep(1)
-            with process(sender_command):
-                time.sleep(1)
+        listener_command = prepare_listen_command(
+            TEST_PORT, receive_directory, scp_ae_title
+        )
+        sender_command = prepare_send_command(TEST_PORT, scp_ae_title, send_file)
+
+        with process(
+            listener_command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT
+        ) as listener_process:
+            stream_output = b""
+            for b in iter(lambda: listener_process.stdout.read(1), b""):
+                stream_output += b
+                if b"Listener Ready" in stream_output:
+                    break
```

Comment:
Given this pattern is re-used a few times, this should probably be spun out into its own function... To be able to still utilise the "with logic" just with your additions it might be worth making a context manager for this section of code. There is a neat decorator in the Python standard library that can convert a defined function generator with a try finally block into a context manager.

I'm on sick leave at the moment, so I'm just on me phone and didn't get a chance to read the rest of your code.

Reply:
Thanks @SimonBiggs, I will look into that suggestion.

But in the meantime, rest up and get well soon. We'll follow up on the remainder of this when you're back in action.

Cheers
Phil